File: src/write_pdf_song.ml
open ExtList
open ExtString
open Printf
open Read_util
open Write_util


module D = Datamodel

let (//) = Filename.concat

let write_mp3 (pf : ('a, unit, string, unit) format4 -> 'a) title = (
  (* pf "mp3 file : %s\n" title  *)
) ;;


let tex_of_string c = (
  List.fold_left ( fun c (sub,by) ->
    let reg = Str.regexp (Str.quote sub) in
      Str.global_replace reg by c
  ) c  [
    "#","$\\sharp$" ;
  ]
)

let write_lyrics fout l = (
  let pf fs = ksprintf ( fun s -> fprintf fout "%s" s) fs in
  let (nbcols,title,l) = (l.D.Lyrics.nb_cols,l.D.Lyrics.title,l.D.Lyrics.data) in
  let () = pf "\\begin{lyricsfont}\n" in 
  let () = pf "\\begin{verse}\n" in
  let () = pf "{\\commentfont \\hl{%s}} \n" (tex_of_string title) in  
  (* let () = pf "\\paragraph{\\commentfont \\hl{%s}} \n" title in  *)
  let () = List.iter ( fun line ->
    if line="\\" then (
      pf "%s" "\n\\end{verse}\n\\begin{verse}\n" 
    )  else (

      let reg = Str.regexp "{\\(.*\\)}" in
      let line = Str.global_replace reg "{\\sethlcolor{grey8}\\commentfont \\hl{\\1}}" line in


      let reg = Str.regexp "\\(#a\\)\\(.*\\)\\(#\\)" in
      let line = Str.global_replace reg "{\\sethlcolor{colora}\\commentafont \\hl{\\2}}" line in



      let rec r line = (
	try
	  let reg = Pcre.regexp "\\[(.*?);(.*?)\\]" in 
	  let s = Pcre.exec ~rex:reg ~pos:0 line in
	  let chord = 
	    let l = Read_util.position_and_silence_or_chord_of_string (Pcre.get_substring s 1) in
	      tex_of_chord l
	  in
	  let word = Pcre.get_substring s 2 in
	  let templ = sprintf "\\textsuperscript{\\textcolor{red}{%s}}\\underline{%s}" chord word in 
	  let line = Pcre.qreplace_first ~rex:reg ~templ:templ line in 
	  r line
	with
	| Not_found -> line
	| Pcre.Error e -> printf "pcre error\n" ; line
      ) in
      let line = r line in

      (* let line = tex_of_string line in  *)
      pf "%s\\\\\n" line
    )
  ) l in
  let () = pf "\\end{verse}\n" in
  let () = pf "\\end{lyricsfont}\n" in 
  let () = pf "\n" in
  ()
) 
let write_grille  song fout g name count = (
  let pf fs = ksprintf ( fun s -> fprintf fout "%s" s) fs in

  let title = g.D.Grille.titre in
  let () = pf "
\\begin{tabular}{c}
\\\\
{\\commentfont \\hl{%s}}
\\\\
" (tex_of_string title) in

  let () = Write_mp_grille.write_mp song name g count in
    pf "
\\includegraphics{%s-grille-%d.mps}
\\end{tabular}
" (Filename.basename name) count ;
    ()
)

let write_tab  song fout tab name count = (
  let pf fs = ksprintf ( fun s -> fprintf fout "%s" s) fs in

  let title = tab.D.Tablature.titre in
    
  let () = pf "
\\begin{tabular}{c}
\\\\
" in 
  let () = if title<>"" then pf"
{\\commentfont \\hl{%s}}
\\\\
" (tex_of_string title) else () in

  let () = Write_mp_tab.write_mp song name tab count in
(*
  let () = List.iter ( fun line ->
    List.iter ( fun bar ->
      List.iter ( fun note ->
	(* pf "%d %d %d" note.D.Tablature.position note.D.Tablature.frette note.D.Tablature.corde *)
	pf " "
      ) bar
    ) line
  ) tab.D.Tablature.lines
  in
*)
    pf "
\\includegraphics{%s-tab-%d.mps}
\\end{tabular}
" (Filename.basename name) count ;
    ()
)

let write_accords  fout l = (
)


let write_song_body fout song = (

  let pf fs = ksprintf ( fun s -> fprintf fout "%s" s) fs in

  let () = List.iter ( fun c ->
    match c with 
    | D.Normal s -> pf "%s\n" s
    | D.Titre _ 
    | D.Auteur _ -> ()
    | D.Grille g -> () (*write_grille ~transpose:song.Song.transpose fout name (g:Accord.t list list list)*)
    | D.Lyrics l -> () (* write_lyrics fout l *)
    | D.Mp3 l -> write_mp3 pf l
    | D.Tab tab -> () (* write_tab fout tab *)
    | D.Accords l -> write_accords fout l
    | D.Transpose i -> ()
    | D.Nb_croches _ -> ()
    | D.PageBreak -> pf "\\newpage\n"
  ) song.D.Song.data in


  let (grilles:D.Grille.t list) = List.rev (List.fold_left ( fun acc c -> match c with | D.Grille g -> g::acc | _ -> acc ) [] song.D.Song.data) in
  let () = pf "\\begin{multicols}{2}\n" in
  let name = Filename.chop_extension song.D.Song.filename in
  let () = List.iteri ( fun index g -> write_grille song fout g name index ) grilles in
  let () = pf "\\end{multicols}\n" in

  let (tabs:D.Tablature.t list) = List.rev (List.fold_left ( fun acc c -> match c with | D.Tab g -> g::acc | _ -> acc ) [] song.D.Song.data) in
  let () = 
    let () = pf "\\begin{multicols}{2}\n" in
    let name = Filename.chop_extension song.D.Song.filename in
    let () = List.iteri ( fun index tab -> write_tab song fout tab name index ) tabs in 
    let () = pf "\\end{multicols}\n" in
      ()
  in

  let lyrics = List.rev (List.fold_left ( fun acc c -> match c with | D.Lyrics l -> l::acc | _ -> acc ) [] song.D.Song.data) in
  let nbcols = List.fold_left ( fun acc l -> if l.D.Lyrics.nb_cols>acc then l.D.Lyrics.nb_cols else acc) 1 lyrics in
  let () = if nbcols>1 then pf "\\begin{multicols}{%d}\n" nbcols in
  let () = List.iter ( fun l -> write_lyrics fout l ) lyrics in
  let () = if nbcols>1 then pf "\\end{multicols}\n" in


  ()
)


let write_preamble fout  = (

  let pf fs = ksprintf ( fun s -> fprintf fout "%s" s) fs in

    
  let _ = pf  "\
\\documentclass[a4paper,portrait]{article}
\\usepackage[utf8]{inputenc}
\\usepackage[T1]{fontenc}
%%\\usepackage{wasysym}
\\usepackage{soulutf8}
\\usepackage{fixltx2e}
%%\\usepackage[pdftex,colorlinks=true,urlcolor=blue,linkcolor=blue]{hyperref}
\\usepackage{hyperref}
\\usepackage{diagbox}
%%\\usepackage{caption}
%%\\usepackage[scale=0.8]{geometry}
\\usepackage[a4paper,inner=1cm,outer=1cm,top=2cm,bottom=2cm]{geometry} 
\\usepackage{color,soul}
\\definecolor{grey}{rgb}{0.7,0.7,0.7}
\\definecolor{grey8}{rgb}{0.8,0.8,0.8}
\\definecolor{colora}{rgb}{0.9,0.7,0.9}
\\sethlcolor{grey8}
\\usepackage[pdftex]{graphicx}
%%\\usepackage{lmodern}
%%\\usepackage{textcomp}
%%\\usepackage{kpfonts}
\\usepackage{array,multirow}
\\usepackage{fancyhdr} 
\\pagestyle{fancy} 
\\usepackage{makeidx}
\\newcolumntype{M}[1]{>{\\centering\\arraybackslash}m{#1}}
%%\\newcolumntype{M}[1]{>{\\centering}m{#1}}
\\newcolumntype{N}{@{}m{0pt}@{}}
\\renewcommand{\\arraystretch}{1.2}
\\usepackage{multicol}
%%\\usepackage{french}
\\setlength{\\columnseprule}{1pt}
\\setlength{\\columnsep}{0cm}
\\usepackage{lastpage}
\\usepackage{needspace}
\\usepackage{titletoc}
\\hypersetup{ pdftitle={}, pdfauthor={},bookmarks=true, bookmarksopen=true,pdftoolbar=true, pdffitwindow=false,colorlinks=false,linkcolor=red, citecolor=red,filecolor=magenta,urlcolor=black }
%%\\usepackage{bookmark}
\\usepackage[metapost,truebbox,mplabels]{mfpic}

%% de la doc de multicols
%%\\setlength{\\textwidth}{39pc}
%%\\setlength{\\textheight}{54pc}
%%\\setlength{\\parindent}{1em}
%%\\setlength{\\parskip}{0pt plus 1pt}
%%\\setlength{\\oddsidemargin}{0pc}
%%\\setlength{\\marginparwidth}{0pc}
%%\\setlength{\\topmargin}{-2.5pc}
%%\\setlength{\\headsep}{20pt}

" in

(*
  let () = pf "\
\\fontencoding{T1}
\\fontfamily{garamond}
\\fontseries{m}
\\fontshape{it}
\\fontsize{22}{25}
\\selectfont
" 
in
*)
  
  let () = pf "
  \\newcommand\\invisiblesection[1]{%%
  \\refstepcounter{section}%%
  \\addcontentsline{toc}{section}{\\protect\\numberline{\\thesection}#1}%%
  \\sectionmark{#1}} \n" in
  let () = pf "\\newcommand*{\\authorfont}{\\fontfamily{ptm}\\fontsize{20}{25}\\fontshape{it}\\selectfont} \n" in
  let () = pf "\\newcommand*{\\titlefont}{\\fontfamily{ptm}\\fontsize{30}{35}\\fontshape{it}\\selectfont} \n" in
  let () = pf "\\newcommand*{\\commentfont}{\\fontfamily{ptm}\\fontsize{12}{15}\\fontshape{it}\\selectfont} \n" in
  let () = pf "\\newcommand*{\\commentafont}{\\fontfamily{ptm}\\fontsize{12}{15}\\fontshape{it}\\selectfont} \n" in
  let () = pf "\\newcommand*{\\lyricstitlefont}{\\color{black}\\fontfamily{ptm}\\fontsize{15}{15}\\fontshape{it}\\selectfont} \n" in
  let () = pf "\\newenvironment{lyricsfont}{\\fontfamily{ptm}\\fontsize{12}{12}\\selectfont}{} \n" in
  let () = pf "\\newenvironment{grillefont}{\\color{blue}\\fontfamily{ptm}\\fontsize{15}{15}\\selectfont}{} \n" in 
  let () = pf "\\newcommand*{\\tabbox}[2][t]{%%
    \\vspace{0pt}\\parbox[#1][3.7\\baselineskip]{0.5cm}{\\strut#2\\strut}}\n" in
  (* let () = pf "\\newcommand{\\tstamp}{\\today}   \n" in *)
  (* let () = pf "\\fancyhead[C]{{\\titlefont %s} \\textsubscript{\\authorfont (%s)}} \n"  song.Song.titre song.Song.auteur in 
  let () = pf "\\fancyhead[R]{} \n" in
  let () = pf "\\fancyhead[L]{} \n" in
  *)
  
  let tm = Unix.localtime (Unix.time ()) in
  let () = pf "\\fancyfoot[L]{généré le %02d/%02d/%04d} \n" (tm.Unix.tm_mday)  (tm.Unix.tm_mon+1) (tm.Unix.tm_year+1900) in
  let () = pf "\\fancyfoot[C]{} \n" in
  let () = pf "\\fancyfoot[R]{page \\thepage~sur \\pageref{LastPage}} \n" in
  let () = pf "\\renewcommand\\headrulewidth{0.02in}" in 
  let () = pf "\\renewcommand\\footrulewidth{0.02in}" in 


  let () = pf "\\Needspace{5\\baselineskip}\n" in

  let () = pf "\\def\\mystrut(#1,#2){\\vrule height #1 depth #2 width 0pt}\n" in
  (* let () = pf "\\newcolumntype{C}{>{\\lower 2ex\\hbox to 6ex\\bgroup\\hss}c<{\\hss\\egroup}}\n" in *)
  let () = pf "\\newcolumntype{C}[1]{%%
   >{\\mystrut(10ex,10ex)\\centering}%%
   p{#1}%%
   <{}}  \n" in
  ()
)


let write_song fout song = (

  let pf fs = ksprintf ( fun s -> fprintf fout "%s" s) fs in

  let () = write_preamble fout in
  let _ = pf  "\
\\title{%s}
" song.D.Song.titre 
  in
  let () = pf "\\author{%s}\n" song.D.Song.auteur in

  let () = pf "\\fancyhead[L]{{\\titlefont %s} } \n"  song.D.Song.titre in 
  let () = pf "\\fancyhead[R]{{\\authorfont %s}} \n"  song.D.Song.auteur in
  let () = pf "\\fancyhead[C]{} \n" in

  let () = pf "
\\begin{document}

" 
  in
  
  let () = write_song_body fout song in
  
  let () = pf "

\\end{document}
"
  in
  ()
)


let write_book ~book   = (
  let fout = open_out "book" ( ("book-"^(Filename.chop_suffix (Filename.basename book.D.Book.filename) ".book")^".tex")) in
  let pf fs = ksprintf ( fun s -> fprintf fout "%s" s) fs in
    
  let () = write_preamble fout in
  let _ = pf  "\
\\title{%s}
" (Filename.basename book.D.Book.titre)
  in

  let () = pf "

\\makeatletter
\\let\\latexl@section\\l@section
\\def\\l@section#1#2{\\begingroup\\let\\numberline\\@gobble\\latexl@section{#1}{#2}\\endgroup}
\\makeatother
\\makeindex

\\begin{document} 
\\maketitle
%%\\begin{multicols}{2}
" in
  let () = pf "
%%\\end{multicolumns}
%%\\titlecontents{section}[0em]
%%{\\vskip 0.5ex}%%
%%{}%% numbered sections formatting
%%{}%% unnumbered sections formatting
%%{}%%

\\begin{multicols}{4}
\\tableofcontents{}
\\end{multicols}
" in
  let () = if book.D.Book.print_index then 
    pf "\\printindex\n" else 
      pf "%%printindex was not set\n"
  in
  let () = List.iter ( fun song ->
    match song with
    | D.Book.S song -> (
      let () = pf "\\clearpage\n" in
      let () = pf "\\index{%s!%s}" song.D.Song.auteur song.D.Song.titre in
      let () = pf "%%\\pdfbookmark[1]{%s}{%s}\n" song.D.Song.titre song.D.Song.titre in
      let () = pf "%%\\invisiblesection{%s}\n" song.D.Song.titre in
      let () = pf "\\fancyhead[L]{{\\invisiblesection{%s (%s)} \\titlefont %s} } \n" 
	song.D.Song.titre song.D.Song.auteur
	song.D.Song.titre in 
      let () = pf "\\fancyhead[R]{{\\authorfont %s}} \n"  song.D.Song.auteur in
      let () = pf "\\fancyhead[C]{} \n" in
      write_song_body fout song
    )
    | D.Book.NF filename -> (
      let () = pf "\\clearpage\n" in 

      let () = pf "\\section{%s (non trouvé)}\n" filename in
      let () = pf "\\fancyhead[L]{{\\titlefont %s} } \n"  filename in
      let () = pf "\\fancyhead[R]{{\\authorfont %s}} \n"  "" in 
      let () = pf "\\fancyhead[C]{} \n" in

      let () = pf "no such song : '%s'\n" filename in
      ()
    )
  ) book.D.Book.songs in

  let () = pf "
\\end{document}
" in
  let () = close_out fout in
  ()



)

File: src/write_pdf_song.cpp
#include <iostream>
#include <vector>
#include <assert.h>
#include <sstream>
#include <fstream>
#include <functional>
#include <ctime>
#include <iomanip>
#include <numeric>

#include "song.h"
#include "datamodel.h"
#include "read_util.h"

extern Datamodel::Accord chord_of_string(const std::string&) ;

std::string replace_chord(const std::string& line) {
  std::function<void (std::string& line)> r = 
    [&r](std::string& line) {
    auto pos1 = line.find('[') ;
    if (pos1==std::string::npos) return ;
    auto pos2 = line.find(']') ;
    if (pos2==std::string::npos) return ;
    
    std::string before = line.substr(0,pos1) ;
    std::string after = line.substr(pos2+1,line.size()-pos2-1) ;
    std::string chord_and_word = line.substr(pos1+1,pos2-pos1-1) ;

    auto pos3 = chord_and_word.find(';') ;
    std::string chord = chord_and_word.substr(0,pos3) ;
    std::string word = chord_and_word.substr(pos3+1,chord_and_word.size()-pos3-1) ;

    std::cout << "before '" << before << "'" << std::endl; 
    std::cout << "chord_and_word '" << chord_and_word << "'" << std::endl; 
    std::cout << "chord '" << chord << "'" << std::endl; 

    Datamodel::Accord a(chord_of_string(chord)) ;
    {
      std::ostringstream oss ;
      std::cout << "note : '" << a.t_.chord_.note_ << "'" << std::endl ;
      oss << "\\textsuperscript{\\textcolor{red}{" << a.t_.chord_.note_ ;
      switch (a.t_.chord_.alteration_) {
      case Datamodel::Accord::None : break ;
      case Datamodel::Accord::Flat : oss << "$\\flat$" ; break ;
      case Datamodel::Accord::Sharp : oss << "$\\sharp$" ; break ;
      default : assert(false) ;
      }
      std::string subscript ;
      if (a.t_.chord_.minor_) { subscript="m" ; }
      if (a.t_.chord_.minor7_) { subscript += "7" ; }
      if (a.t_.chord_.sus4_) { subscript += "4" ; } 
      if (a.t_.chord_.major7_) { subscript += "7M" ; }
      if (subscript != "" ) { oss << "\\textsubscript{\\small{" << subscript << "}}" ; }
      oss << "}}" ;
      chord = oss.str() ;
    }

    std::cout << "word '" << word << "'" << std::endl; 
    std::cout << "after '" << after << "'" << std::endl; 
    std::string line2(before + chord + "\\underline{" + word + "}" + after) ;
    std::cout << "line2 '" << line2 << "'" << std::endl; 
    line=line2 ;
    // std::string line2(after) ;
    // assert(line.size() > line2.size()+2) ;
    r(line) ;
  } ;
  std::string ret(line) ;
  std::cout << "BBBBBBBBBBefore : " << line << std::endl ;
  r(ret) ;
  std::cout << "AAAAAAAAAAAAAafter : " << ret << std::endl ;
  return ret ;
}
  


std::string mp_grille_filename (const Datamodel::Conf& conf,const Song& song,int count) {
  std::string ret(replace_extension(song.filename_,"")) ;
  ret = replace_path(ret,conf.srcdir_.c_str(),conf.builddir_.c_str()) ;
  replace_extension(ret,"") ;
  ret += "-grille-" + std::to_string(count) + ".mp" ;
  return ret ;
}

std::string mp_tab_filename (const Datamodel::Conf& conf,const Song& song,int count) {
  std::string ret(replace_extension(song.filename_,"")) ;
  ret = replace_path(ret,conf.srcdir_.c_str(),conf.builddir_.c_str()) ;
  replace_extension(ret,"") ;
  ret += "-tab-" + std::to_string(count) + ".mp" ;
  return ret ;
}

extern void grille_write_mp(const std::string&filename,const Grille& grille) ;

extern void tab_write_mp(const Song& song,const std::string&filename,const Tablature& tab) ;


void write_preamble(std::ofstream& fout) {

  fout << "\
\\documentclass[a4paper,portrait]{article}\n\
\\usepackage[utf8]{inputenc}\n\
\\usepackage[T1]{fontenc}\n\
%\\usepackage{wasysym}\n\
\\usepackage{soulutf8}\n\
\\usepackage{fixltx2e}\n\
%\\usepackage[pdftex,colorlinks=true,urlcolor=blue,linkcolor=blue]{hyperref}\n\
\\usepackage{hyperref}\n\
\\usepackage{diagbox}\n\
%\\usepackage{caption}\n\
%\\usepackage[scale=0.8]{geometry}\n\
\\usepackage[a4paper,inner=1cm,outer=1cm,top=2cm,bottom=2cm]{geometry} \n\
\\usepackage{color,soul}\n\
\\definecolor{grey}{rgb}{0.7,0.7,0.7}\n\
\\definecolor{grey8}{rgb}{0.8,0.8,0.8}\n\
\\definecolor{colora}{rgb}{0.9,0.7,0.9}\n\
\\sethlcolor{grey8}\n\
\\usepackage[pdftex]{graphicx}\n\
%\\usepackage{lmodern}\n\
%\\usepackage{textcomp}\n\
%\\usepackage{kpfonts}\n\
\\usepackage{array,multirow}\n\
\\usepackage{fancyhdr} \n\
\\pagestyle{fancy} \n\
\\usepackage{makeidx}\n\
\\newcolumntype{M}[1]{>{\\centering\\arraybackslash}m{#1}}\n\
%\\newcolumntype{M}[1]{>{\\centering}m{#1}}\n\
\\newcolumntype{N}{@{}m{0pt}@{}}\n\
\\renewcommand{\\arraystretch}{1.2}\n\
\\usepackage{multicol}\n\
%\\usepackage{french}\n\
\\setlength{\\columnseprule}{1pt}\n\
\\setlength{\\columnsep}{0cm}\n\
\\usepackage{lastpage}\n\
\\usepackage{needspace}\n\
\\usepackage{titletoc}\n\
\\hypersetup{ pdftitle={}, pdfauthor={},bookmarks=true, bookmarksopen=true,pdftoolbar=true, pdffitwindow=false,colorlinks=false,linkcolor=red, citecolor=red,filecolor=magenta,urlcolor=black }\n\
%\\usepackage{bookmark}\n\
\\usepackage[metapost,truebbox,mplabels]{mfpic}\n\
\n\
% de la doc de multicols\n\
%\\setlength{\\textwidth}{39pc}\n\
%\\setlength{\\textheight}{54pc}\n\
%\\setlength{\\parindent}{1em}\n\
%\\setlength{\\parskip}{0pt plus 1pt}\n\
%\\setlength{\\oddsidemargin}{0pc}\n\
%\\setlength{\\marginparwidth}{0pc}\n\
%\\setlength{\\topmargin}{-2.5pc}\n\
%\\setlength{\\headsep}{20pt}\n\
\n\
\n\
  \\newcommand\\invisiblesection[1]{%\n\
  \\refstepcounter{section}%\n\
  \\addcontentsline{toc}{section}{\\protect\\numberline{\\thesection}#1}%\n\
  \\sectionmark{#1}} \n\
\\newcommand*{\\authorfont}{\\fontfamily{ptm}\\fontsize{20}{25}\\fontshape{it}\\selectfont} \n\
\\newcommand*{\\titlefont}{\\fontfamily{ptm}\\fontsize{30}{35}\\fontshape{it}\\selectfont} \n\
\\newcommand*{\\commentfont}{\\fontfamily{ptm}\\fontsize{12}{15}\\fontshape{it}\\selectfont} \n\
\\newcommand*{\\commentafont}{\\fontfamily{ptm}\\fontsize{12}{15}\\fontshape{it}\\selectfont} \n\
\\newcommand*{\\lyricstitlefont}{\\color{black}\\fontfamily{ptm}\\fontsize{15}{15}\\fontshape{it}\\selectfont} \n\
\\newenvironment{lyricsfont}{\\fontfamily{ptm}\\fontsize{12}{12}\\selectfont}{} \n\
\\newenvironment{grillefont}{\\color{blue}\\fontfamily{ptm}\\fontsize{15}{15}\\selectfont}{} \n\
\\newcommand*{\\tabbox}[2][t]{%\n\
    \\vspace{0pt}\\parbox[#1][3.7\\baselineskip]{0.5cm}{\\strut#2\\strut}}\n\
" ;

  std::time_t t = std::time(nullptr) ;
  std::tm* tm = std::localtime(&t) ;
  fout << "\\fancyfoot[L]{généré le " 
       << std::setw(2) << std::setfill('0') << tm->tm_mday << "/" 
       << std::setw(2) << std::setfill('0') << (tm->tm_mon+1) << "/" 
       << std::setw(4) << (tm->tm_year+1900) << "} \n\
" ;

  fout << "\\fancyfoot[C]{} \n\
\\fancyfoot[R]{page \\thepage~sur \\pageref{LastPage}} \n\
\\renewcommand\\headrulewidth{0.02in}\\renewcommand\\footrulewidth{0.02in}\
\\Needspace{5\\baselineskip}\n\
\\def\\mystrut(#1,#2){\\vrule height #1 depth #2 width 0pt}\n\
\\newcolumntype{C}[1]{%\n\
   >{\\mystrut(10ex,10ex)\\centering}%\n\
   p{#1}%\n\
   <{}}  \n\
" ;
}

const void Song::write(const Datamodel::Conf& la_conf) {

  std::cout << "Song::write " << filename_ << std::endl ;


  std::accumulate(
		  grilles_.begin(),grilles_.end(),0,
		  [this,&la_conf](int index,const Grille& g) {
		    grille_write_mp(mp_grille_filename(la_conf,*this,index),g) ;
		    return index+1 ;
		  }) ;

  std::accumulate(
		  tablatures_.begin(),tablatures_.end(),0,
		  [this,&la_conf](int index,const Tablature& t) {
		    tab_write_mp(*this,mp_tab_filename(la_conf,*this,index),t) ;
		    return index+1 ;
		  }) ;


  std::function<void(std::ofstream&)> write_1 =
    [this](std::ofstream&fout) {
    fout << "\
\\title{" << this->titre_ << "}\n\
\\author{" << this->auteur_ << "}\n\
\\fancyhead[L]{{\\titlefont " << this->titre_ << "} } \n\
\\fancyhead[R]{{\\authorfont "<<  this->auteur_ << "}} \n\
\\fancyhead[C]{} \n\
\n\
\\begin{document}\n\
\n\
"
    ;
  } ;



  std::function<void(std::ofstream&,const std::string&)>write_grilles =
    [this](std::ofstream&fout,const std::string &file_basename) {
    int count=0 ;
    fout << "\
\\begin{multicols}{2}\n\
\n\
" ;
    for (auto g:this->grilles_) {
      fout << "\
\\begin{tabular}{c}\n\
\\\\\n\
{\\commentfont \\hl{" << tex_of_string(g.t_.titre_) << "}}\n\
\\\\\n\
\n\
\\includegraphics{" << basename(file_basename) << "-grille-" << count << ".mps}\n\
\\end{tabular}\n\
" ;
      count++ ;
    }
    fout << "\
\\end{multicols}\n\
" ;
  } ;


  std::function<void(std::ofstream&,const std::string&)>write_tablatures =
    [this](std::ofstream&fout,const std::string &file_basename) {
    int count=0 ;
    fout << "\\begin{multicols}{2}\n" ;
    for (auto b:this->tablatures_) {
      fout << "\
\\begin{tabular}{c}\n\
\\\\\n\n\
{\\commentfont \\hl{" << tex_of_string(b.t_.titre_) << "}}\n\
\\\\\n\
\n\
\\includegraphics{" << basename(file_basename) << "-tab-" << count << ".mps}\n \
\\end{tabular}\n\
" ;
      count++ ;
    }
    fout << "\
\\end{multicols}\n\
" ;
  } ;


  std::function<void(std::ofstream&)>write_lyrics =
    [this](std::ofstream&fout) {
    // fout << "LLLLLLLLLLLLLL" << this->lyrics_.size() << std::endl; 
    fout << "\\begin{multicols}{2}\n" ;
    for (auto l:this->lyrics_) {
      fout << "\\begin{lyricsfont}\n" ;
      fout << "\\begin{verse}\n" ;
      fout << "{\\commentfont \\hl{" << tex_of_string(l.title_) << "}} \n" ;
      for (auto line:l.data_) {
	if (line=="\\") {
	  fout << "\n\\end{verse}\n\\begin{verse}\n" ;
	}
	else if (line.size()>2 && line[0]=='{') {
	  fout << "{\\sethlcolor{grey8}\\commentfont \\hl" << line << "}\\\\\n" ;
	}
	else {
	  fout << tex_of_string(replace_chord(line)) << "\\\\" << std::endl; 
	}
      }
      fout << "\\end{verse}\n" ;
      fout << "\\end{lyricsfont}\n\n" ;
    }
    fout << "\\end{multicols}\n\n" ;
  } ;

  std::string path (replace_extension(filename_,".tex")) ;
  path = replace_path(path,la_conf.srcdir_.c_str(),la_conf.builddir_.c_str()) ;
  std::cout << "Write tex file '" << path << "'" << std::endl ;
  std::ofstream fout(path) ;
  write_preamble(fout) ;
  write_1(fout) ;
  write_grilles(fout,replace_extension(path,"")) ;
  write_tablatures(fout,replace_extension(path,"")) ;
  write_lyrics(fout) ;
  fout << "\n\\end{document}\n" ;
}
